subsys/sensing/sensor_mgmt.c

```diff
@@ -0,0 +1,329 @@
+	int i;
+
+	LOG_INF("sensing init begin...");
+
+	if (ctx->sensing_initialized) {
+		LOG_INF("sensing is already initialized");
+		return 0;
+	}
+
+	if (ctx->sensor_num == 0) {
+		LOG_WRN("no sensor created by device tree yet");
+		return 0;
+	}
+
+	for (i = 0; i < ctx->sensor_num; i++) {
+		sensor = create_sensor(*dt_list++);
+		if (!sensor) {
+			LOG_ERR("sensing init, create sensor error");
+			return -EINVAL;
+		}
+		ctx->sensors[i] = sensor;
+	}
+
+	sort_sensors(ctx);
+
+	for_each_sensor(ctx, i, sensor) {
+		ret = init_sensor(sensor, sensor->dt->reporter_num);
+		if (ret) {
+			LOG_ERR("sensor:%s initial error", sensor->dt->dev->name);
+		}
+		state = (ret ? SENSING_SENSOR_STATE_OFFLINE : SENSING_SENSOR_STATE_READY);
+		ret = set_sensor_state(sensor, state);
+		if (ret) {
+			LOG_ERR("set sensor:%s state:%d error", sensor->dt->dev->name, state);
+		}
+		LOG_INF("sensing init, sensor:%s ret:%d", sensor->dt->dev->name, ret);
+	}
+
+	return ret;
+}
+
+int open_sensor(struct sensing_sensor *sensor, sensing_sensor_handle_t *handle)
+{
+	struct sensing_connection *conn;
+
+	*handle = NULL;
+	/* allocate struct sensing_connection *conn and conn data for application client */
+	conn = malloc(sizeof(*conn) + sensor->sample_size);
+	if (!conn) {
+		LOG_ERR("malloc memory for struct sensing_connection error");
+		return -ENOMEM;
+	}
+	conn->data = (uint8_t *)conn + sizeof(*conn);
+
+	/* create connection from sensor to application(client = NULL) */
+	init_connection(conn, sensor, NULL);
+
+	*handle = conn;
+
+	return 0;
+}
+
+int close_sensor(struct sensing_connection *conn)
+{
+	__ASSERT(conn && conn->source, "connection or reporter not be NULL");
+
+	__ASSERT(!conn->sink, "sensor derived from device tree cannot be closed");
+
+	sys_slist_find_and_remove(&conn->source->client_list, &conn->snode);
+
+	free(conn);
+
+	return 0;
+}
+
+int sensing_register_callback(struct sensing_connection *conn,
+			      const struct sensing_callback_list *cb_list)
+{
+	__ASSERT(conn, "register sensing callback list, connection not be NULL");
+
+	__ASSERT(!conn->sink, "only connection to application could register sensing callback");
+
+	__ASSERT(cb_list, "callback should not be NULL");
+
+	conn->data_evt_cb = cb_list->on_data_event;
+
+	return 0;
+}
+
+int set_interval(struct sensing_connection *conn, uint32_t interval)
+{
+	return -ENOTSUP;
+}
+
+int get_interval(struct sensing_connection *conn, uint32_t *interval)
+{
+	return -ENOTSUP;
+}
+
+int set_sensitivity(struct sensing_connection *conn, int8_t index, uint32_t sensitivity)
+{
+	return -ENOTSUP;
+}
+
+int get_sensitivity(struct sensing_connection *conn, int8_t index, uint32_t *sensitivity)
+{
+	return -ENOTSUP;
+}
+
+int sensing_get_sensors(int *sensor_nums, const struct sensing_sensor_info **info)
+{
+	int i;
+
+	__ASSERT(info, "sensing get sensors, sensor info not be NULL");
+
+	*sensor_nums = ARRAY_SIZE(sensors_dt_list);
+	for (i = 0; i < *sensor_nums; i++) {
+		info[i] = &sensors_dt_list[i]->info;
```

Comment:
Last change needed when deleting SENSING_DT_ITEM.
```suggestion
		info[i] = &sensors_dt_list[i].info;
```